File: mettle/backend/app/schemas/candidate.py
from pydantic import BaseModel, EmailStr
from typing import Optional, List, Literal, Any
from uuid import UUID
from datetime import datetime


SourceType = Literal["LinkedIn", "GitHub", "Referral", "CareerPage", "Indeed"]
CandidateStatus = Literal["New", "Screening", "Interview", "Offer", "Hired", "Rejected"]


class WorkExperience(BaseModel):
    """Work experience entry."""
    id: str
    title: str
    company: str
    start_date: str
    end_date: Optional[str] = None
    location: Optional[str] = None
    description: Optional[str] = None


class Education(BaseModel):
    """Education entry."""
    id: str
    school: str
    degree: str
    start_date: str
    end_date: Optional[str] = None
    grade: Optional[str] = None


class Certification(BaseModel):
    """Certification entry."""
    id: str
    name: str
    issuer: str
    issue_date: str
    credential_id: Optional[str] = None


class CandidateBase(BaseModel):
    """Base candidate schema."""
    name: str
    email: EmailStr
    phone: Optional[str] = None
    photo_url: Optional[str] = None
    role: str
    source: SourceType
    location: Optional[str] = None
    skills: Optional[List[str]] = None
    summary: Optional[str] = None
    experience: Optional[List[WorkExperience]] = None
    education: Optional[List[Education]] = None
    certifications: Optional[List[Certification]] = None


class CandidateCreate(CandidateBase):
    """Schema for creating a candidate."""
    applied_job_id: Optional[UUID] = None


class CandidateUpdate(BaseModel):
    """Schema for updating a candidate."""
    name: Optional[str] = None
    email: Optional[EmailStr] = None
    phone: Optional[str] = None
    photo_url: Optional[str] = None
    role: Optional[str] = None
    source: Optional[SourceType] = None
    status: Optional[CandidateStatus] = None
    score: Optional[int] = None
    location: Optional[str] = None
    skills: Optional[List[str]] = None
    tags: Optional[List[str]] = None
    summary: Optional[str] = None
    resume_url: Optional[str] = None
    experience: Optional[List[WorkExperience]] = None
    education: Optional[List[Education]] = None
    certifications: Optional[List[Certification]] = None


class CandidateResponse(CandidateBase):
    """Candidate response schema."""
    id: UUID
    status: CandidateStatus
    score: int
    tags: Optional[List[str]] = None
    experience_years: int
    resume_url: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True

File: mettle/frontend/apps/web/src/lib/validations/candidate.ts
import { z } from 'zod';

// Candidate basic info schema
export const candidateBasicInfoSchema = z.object({
    firstName: z.string().min(1, 'First name is required'),
    lastName: z.string().min(1, 'Last name is required'),
    email: z.string().email('Invalid email address'),
    phone: z.string().optional(),
    role: z.string().optional(),
    location: z.string().optional(),
    gender: z.string().optional(),
    dob: z.string().optional(),
    linkedin: z.string().url().optional().or(z.literal('')),
    portfolio: z.string().url().optional().or(z.literal('')),
    photoUrl: z.string().optional(),
    appliedJobId: z.string().optional()
});

// Work experience schema
export const workExperienceSchema = z.object({
    id: z.string(),
    title: z.string(),
    company: z.string(),
    type: z.enum(['Full-time', 'Part-time', 'Contract', 'Freelance', 'Internship']),
    location: z.string(),
    startDate: z.string(),
    endDate: z.string(),
    current: z.boolean(),
    description: z.string()
});

// Education schema
export const educationSchema = z.object({
    id: z.string(),
    school: z.string(),
    degree: z.string(),
    field: z.string(),
    startDate: z.string(),
    endDate: z.string(),
    grade: z.string(),
    activities: z.string()
});

// Certification schema
export const certificationSchema = z.object({
    id: z.string(),
    name: z.string(),
    issuer: z.string(),
    issueDate: z.string(),
    expirationDate: z.string(),
    credentialId: z.string()
});

// Full candidate form schema
export const candidateFormSchema = z.object({
    basicInfo: candidateBasicInfoSchema,
    summary: z.string().optional(),
    skills: z.string().optional(),
    experience: z.array(workExperienceSchema),
    education: z.array(educationSchema),
    certifications: z.array(certificationSchema)
});

// Type inference
export type CandidateBasicInfoInput = z.infer<typeof candidateBasicInfoSchema>;
export type WorkExperienceInput = z.infer<typeof workExperienceSchema>;
export type EducationInput = z.infer<typeof educationSchema>;
export type CertificationInput = z.infer<typeof certificationSchema>;
export type CandidateFormInput = z.infer<typeof candidateFormSchema>;
